File: src/main/java/app/tweditor/LoadFile.java
package app.tweditor;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import javax.swing.SwingUtilities;

public class LoadFile extends Thread
{
  private final ProgressDialog progressDialog;
  private final GameSession session;
  private final AppEnvironment environment;
  private File file;
  private boolean loadSuccessful = false;

  public LoadFile(ProgressDialog dialog, GameSession session, AppEnvironment environment, File file)
  {
    this.progressDialog = dialog;
    this.session = session;
    this.environment = environment;
    this.file = file;
  }

  public void run()
  {
    InputStream in = null;
    FileOutputStream out = null;
    try
    {
      SaveDatabase saveDatabase = new SaveDatabase(this.environment, this.file);
      saveDatabase.load();
      this.progressDialog.updateProgress(25);
      String saveName = saveDatabase.getName();
      saveDatabase.setSavePrefix(saveName + this.environment.getFileSeparator());

      int sep = saveName.indexOf(' ');
      if ((sep != 6) || (!Character.isDigit(saveName.charAt(0)))) {
        throw new DBException("Save name is not formatted correctly");
      }
      this.session.setSmmName("save_" + saveName.substring(0, 6) + ".smm");
      SaveEntry saveEntry = saveDatabase.getEntry(this.session.getSmmName());
      if (saveEntry == null) {
        throw new DBException("Save does not contain " + this.session.getSmmName());
      }
      in = saveEntry.getInputStream();
      if (this.session.getSmmFile().exists()) {
        this.session.getSmmFile().delete();
      }
      byte[] buffer = new byte[4096];
      out = new FileOutputStream(this.session.getSmmFile());
      int count;
      while ((count = in.read(buffer)) > 0) {
        out.write(buffer, 0, count);
      }
      in.close();
      in = null;
      out.close();
      out = null;
      Database smmDatabase = new Database(this.environment, this.session.getSmmFile());
      smmDatabase.load();
      this.progressDialog.updateProgress(35);

      DBList list = (DBList)smmDatabase.getTopLevelStruct().getValue();
      String startingMod = list.getString("StartingMod");
      if (startingMod.length() == 0) {
        throw new DBException("StartingMod not found in SMM database");
      }
      DBElement element = list.getElement("QuestBase_list");
      if ((element == null) || (element.getType() != 15)) {
        throw new DBException("QuestBaseList not found in SMM database");
      }
      DBList questList = (DBList)element.getValue();
      if (questList.getElementCount() == 0) {
        throw new DBException("No quest list found in SMM database");
      }
      DBList fieldList = (DBList)questList.getElement(0).getValue();
      String questDBName = fieldList.getString("QuestBase");
      if (questDBName.length() == 0) {
        throw new DBException("No quest database name found in SMM database");
      }

      this.session.setModName(startingMod + ".sav");
      saveEntry = saveDatabase.getEntry(this.session.getModName());
      if (saveEntry == null) {
        throw new DBException("Save does not contain " + this.session.getModName());
      }
      in = saveEntry.getInputStream();
      if (this.session.getModFile().exists()) {
        this.session.getModFile().delete();
      }

      buffer = new byte[4096];
      out = new FileOutputStream(this.session.getModFile());
      while ((count = in.read(buffer)) > 0) {
        out.write(buffer, 0, count);
      }
      in.close();
      in = null;
      out.close();
      out = null;
      this.progressDialog.updateProgress(50);

      ResourceDatabase modDatabase = new ResourceDatabase(this.session.getModFile());
      modDatabase.load();
      this.progressDialog.updateProgress(60);

      ResourceEntry resourceEntry = modDatabase.getEntry("module.ifo");
      if (resourceEntry == null) {
        throw new DBException("Save does not contain module.ifo");
      }
      in = resourceEntry.getInputStream();
      if (this.session.getDatabaseFile().exists()) {
        this.session.getDatabaseFile().delete();
      }
      out = new FileOutputStream(this.session.getDatabaseFile());
      while ((count = in.read(buffer)) > 0) {
        out.write(buffer, 0, count);
      }
      in.close();
      in = null;
      out.close();
      out = null;
      this.progressDialog.updateProgress(75);

      Database database = new Database(this.environment, this.session.getDatabaseFile());
      database.load();
      list = (DBList)database.getTopLevelStruct().getValue();
      element = list.getElement("Mod_PlayerList");
      if ((element == null) || (element.getType() != 15)) {
        throw new DBException("module.ifo does not contain Mod_PlayerList");
      }
      list = (DBList)element.getValue();
      if (list.getElementCount() == 0) {
        throw new DBException("Mod_PlayerList is empty");
      }
      this.progressDialog.updateProgress(80);

      String fileName = questDBName + ".qdb";
      saveEntry = saveDatabase.getEntry(fileName);
      if (saveEntry == null) {
        throw new DBException("Save does not contain " + fileName);
      }
      in = saveEntry.getInputStream();
      Database questDatabase = new Database(this.environment);
      questDatabase.load(in);
      in.close();
      in = null;
      list = (DBList)questDatabase.getTopLevelStruct().getValue();
      element = list.getElement("Quests");
      if ((element == null) || (element.getType() != 15)) {
        throw new DBException("Quests not found in quest database");
      }
      questList = (DBList)element.getValue();
      this.progressDialog.updateProgress(85);

      count = questList.getElementCount();
      List<Quest> quests = new ArrayList<>(count);
      for (int i = 0; i < count; i++) {
        fieldList = (DBList)questList.getElement(i).getValue();
        String resourceName = fieldList.getString("File");
        fileName = resourceName + ".qst";
        saveEntry = saveDatabase.getEntry(fileName);
        if (saveEntry == null) {
          throw new DBException("Save does not contain " + fileName);
        }
        in = saveEntry.getInputStream();
        questDatabase = new Database(this.environment);
        questDatabase.load(in);
        in.close();
        in = null;
        Quest quest = new Quest(resourceName, questDatabase.getTopLevelStruct());
        if (quest.getQuestName().length() > 0) {
          quests.add(quest);
        }
      }
      this.session.setQuests(quests);

      this.session.setPlayerName("player.utc");
      saveEntry = saveDatabase.getEntry(this.session.getPlayerName());
      if (saveEntry == null) {
        throw new DBException("Save does not contain " + this.session.getPlayerName());
      }
      in = saveEntry.getInputStream();
      if (this.session.getPlayerFile().exists()) {
        this.session.getPlayerFile().delete();
      }
      out = new FileOutputStream(this.session.getPlayerFile());
      while ((count = in.read(buffer)) > 0) {
        out.write(buffer, 0, count);
      }
      in.close();
      in = null;
      out.close();
      out = null;

      Database playerDatabase = new Database(this.environment, this.session.getPlayerFile());
      playerDatabase.load();

      this.progressDialog.updateProgress(100);

      this.session.setSaveDatabase(saveDatabase); //.TheWitcherSave
      this.session.setModDatabase(modDatabase); //.sav
      this.session.setDatabase(database); //.sav -> 'module.ifo'
      this.session.setPlayerDatabase(playerDatabase); //player.utc
      this.session.setSmmDatabase(smmDatabase); //.smm
      this.loadSuccessful = true;
    } catch (DBException exc) {
      Main.logException("Save file structure is not valid", exc);
    } catch (IOException exc) {
      Main.logException("Unable to read save file", exc);
    } catch (Throwable exc) {
      Main.logException("Exception while opening save file", exc);
    }

    try
    {
      if (in != null) {
        in.close();
      }
      if (out != null) {
        out.close();
      }

    }
    catch (IOException exc)
    {
    }

    SwingUtilities.invokeLater(new Runnable() {
      public void run() {
        LoadFile.this.progressDialog.closeDialog(LoadFile.this.loadSuccessful);
      }
    });
  }
}

File: src/main/kotlin/app/tweditor/LoadFile.kt
package app.tweditor

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import javax.swing.SwingUtilities

class LoadFile(
    private val progressDialog: ProgressDialog,
    private val session: GameSession,
    private val environment: AppEnvironment,
    private var file: File
) : Thread() {
    private var loadSuccessful = false

    override fun run() {
        var input: InputStream? = null
        var out: FileOutputStream? = null
        try {
            val saveDatabase = SaveDatabase(environment, this.file)
            saveDatabase.load()
            progressDialog.updateProgress(25)
            val saveName = saveDatabase.getName()
            saveDatabase.setSavePrefix(saveName + environment.fileSeparator)

            val sep = saveName.indexOf(' ')
            if (sep != 6 || !Character.isDigit(saveName[0])) {
                throw DBException("Save name is not formatted correctly")
            }
            session.setSmmName("save_" + saveName.substring(0, 6) + ".smm")
            var saveEntry = saveDatabase.getEntry(session.getSmmName()!!)
            if (saveEntry == null) {
                throw DBException("Save does not contain " + session.getSmmName())
            }
            input = saveEntry.getInputStream()
            if (session.smmFile.exists()) {
                session.smmFile.delete()
            }
            var buffer = ByteArray(4096)
            out = FileOutputStream(session.smmFile)
            var count: Int
            while (input!!.read(buffer).also { count = it } > 0) {
                out!!.write(buffer, 0, count)
            }
            input!!.close()
            input = null
            out!!.close()
            out = null
            val smmDatabase = Database(environment, session.smmFile)
            smmDatabase.load()
            progressDialog.updateProgress(35)

            var list = smmDatabase.getTopLevelStruct()!!.getValue() as DBList
            val startingMod = list.getString("StartingMod")
            if (startingMod.isEmpty()) {
                throw DBException("StartingMod not found in SMM database")
            }
            var element = list.getElement("QuestBase_list")
            if (element == null || element.getType() != 15) {
                throw DBException("QuestBaseList not found in SMM database")
            }
            var questList = element.getValue() as DBList
            if (questList.getElementCount() == 0) {
                throw DBException("No quest list found in SMM database")
            }
            var fieldList = questList.getElement(0).getValue() as DBList
            val questDBName = fieldList.getString("QuestBase")
            if (questDBName.isEmpty()) {
                throw DBException("No quest database name found in SMM database")
            }

            session.setModName(startingMod + ".sav")
            saveEntry = saveDatabase.getEntry(session.getModName()!!)
            if (saveEntry == null) {
                throw DBException("Save does not contain " + session.getModName())
            }
            input = saveEntry.getInputStream()
            if (session.modFile.exists()) {
                session.modFile.delete()
            }

            buffer = ByteArray(4096)
            out = FileOutputStream(session.modFile)
            while (input!!.read(buffer).also { count = it } > 0) {
                out!!.write(buffer, 0, count)
            }
            input!!.close()
            input = null
            out!!.close()
            out = null
            progressDialog.updateProgress(50)

            val modDatabase = ResourceDatabase(session.modFile)
            modDatabase.load()
            progressDialog.updateProgress(60)

            val resourceEntry = modDatabase.getEntry("module.ifo")
            if (resourceEntry == null) {
                throw DBException("Save does not contain module.ifo")
            }
            input = resourceEntry.getInputStream()
            if (session.databaseFile.exists()) {
                session.databaseFile.delete()
            }
            out = FileOutputStream(session.databaseFile)
            while (input!!.read(buffer).also { count = it } > 0) {
                out!!.write(buffer, 0, count)
            }
            input!!.close()
            input = null
            out!!.close()
            out = null
            progressDialog.updateProgress(75)

            val database = Database(environment, session.databaseFile)
            database.load()
            list = database.getTopLevelStruct()!!.getValue() as DBList
            element = list.getElement("Mod_PlayerList")
            if (element == null || element.getType() != 15) {
                throw DBException("module.ifo does not contain Mod_PlayerList")
            }
            list = element.getValue() as DBList
            if (list.getElementCount() == 0) {
                throw DBException("Mod_PlayerList is empty")
            }
            progressDialog.updateProgress(80)

            var fileName = questDBName + ".qdb"
            saveEntry = saveDatabase.getEntry(fileName)
            if (saveEntry == null) {
                throw DBException("Save does not contain " + fileName)
            }
            input = saveEntry.getInputStream()
            var questDatabase = Database(environment)
            questDatabase.load(input)
            input!!.close()
            input = null
            list = questDatabase.getTopLevelStruct()!!.getValue() as DBList
            element = list.getElement("Quests")
            if (element == null || element.getType() != 15) {
                throw DBException("Quests not found in quest database")
            }
            questList = element.getValue() as DBList
            progressDialog.updateProgress(85)

            count = questList.getElementCount()
            val quests = ArrayList<Quest>(count)
            for (questElement in questList) {
                fieldList = questElement.getValue() as DBList
                val resourceName = fieldList.getString("File")
                fileName = resourceName + ".qst"
                saveEntry = saveDatabase.getEntry(fileName)
                if (saveEntry == null) {
                    throw DBException("Save does not contain " + fileName)
                }
                input = saveEntry.getInputStream()
                questDatabase = Database(environment)
                questDatabase.load(input)
                input!!.close()
                input = null
                val quest = Quest(resourceName, questDatabase.getTopLevelStruct()!!)
                if (quest.questName.isNotEmpty()) {
                    quests.add(quest)
                }
            }
            session.setQuests(quests)

            session.setPlayerName("player.utc")
            saveEntry = saveDatabase.getEntry(session.getPlayerName()!!)
            if (saveEntry == null) {
                throw DBException("Save does not contain " + session.getPlayerName())
            }
            input = saveEntry.getInputStream()
            if (session.playerFile.exists()) {
                session.playerFile.delete()
            }
            out = FileOutputStream(session.playerFile)
            while (input!!.read(buffer).also { count = it } > 0) {
                out!!.write(buffer, 0, count)
            }
            input!!.close()
            input = null
            out!!.close()
            out = null

            val playerDatabase = Database(environment, session.playerFile)
            playerDatabase.load()

            progressDialog.updateProgress(100)

            session.saveDatabase = saveDatabase
            session.modDatabase = modDatabase
            session.database = database
            session.playerDatabase = playerDatabase
            session.smmDatabase = smmDatabase
            this.loadSuccessful = true
        } catch (exc: DBException) {
            Main.logException("Save file structure is not valid", exc)
        } catch (exc: IOException) {
            Main.logException("Unable to read save file", exc)
        } catch (exc: Throwable) {
            Main.logException("Exception while opening save file", exc)
        }

        try {
            input?.close()
            out?.close()
        } catch (exc: IOException) {
        }

        SwingUtilities.invokeLater {
            progressDialog.closeDialog(loadSuccessful)
        }
    }
}
